pep-0xxx.rst

```diff
@@ -0,0 +1,1149 @@
+implementation.
+
+::
+
+    class TLSVersion(Enum):
+        MINIMUM_SUPPORTED
+        SSLv2
+        SSLv3
+        TLSv1
+        TLSv1_1
+        TLSv1_2
+        TLSv1_3
+        MAXIMUM_SUPPORTED
+
+
+Errors
+~~~~~~
+
+This module would define three base classes for use with error handling. Unlike
+many of the the other classes defined here, these classes are not abstract, as
+they have no behaviour. They exist simply to signal certain common behaviours.
+Backends should subclass these exceptions in their own packages, but needn't
+define any behaviour for them.
+
+In general, concrete implementations should subclass these exceptions rather
+than throw them directly. This makes it moderately easier to determine which
+concrete TLS implementation is in use during debugging of unexpected errors.
+However, this is not mandatory.
+
+The definitions of the errors are below::
+
+    class TLSError(Exception):
+        """
+        The base exception for all TLS related errors from any backend.
+        Catching this error should be sufficient to catch *all* TLS errors,
+        regardless of what backend is used.
+        """
+
+    class WantWriteError(TLSError):
+        """
+        A special signaling exception used only when non-blocking or
+        buffer-only I/O is used. This error signals that the requested
+        operation cannot complete until more data is written to the network,
+        or until the output buffer is drained.
+        """
+
+    class WantReadError(TLSError):
+        """
+        A special signaling exception used only when non-blocking or
+        buffer-only I/O is used. This error signals that the requested
+        operation cannot complete until more data is read from the network, or
+        until more data is available in the input buffer.
+        """
+
+
+Certificates
+~~~~~~~~~~~~
+
+This module would define an abstract X509 certificate class. This class would
+have almost no behaviour, as the goal of this module is not to provide all
+possible relevant cryptographic functionality that could be provided by X509
+certificates. Instead, all we need is the ability to signal the source of a
+certificate to a concrete implementation.
+
+For that reason, this certificate implementation defines only constructors. In
+essence, the certificate object in this module could be as abstract as a handle
+that can be used to locate a specific certificate.
+
+Concrete implementations may choose to provide alternative constructors, e.g.
+to load certificates from HSMs. If a common interface emerges for doing this,
+this module may be updated to provide a standard constructor for this use-case
+as well.
+
+Concrete implementations should aim to have Certificate objects be hashable if
+at all possible. This will help ensure that TLSConfiguration objects used with
+an individual concrete implementation are also hashable.
+
+    class Certificate(metaclass=ABCMeta):
```

Comment:
Take a look at the `TLSConfiguration` object, which you'll see takes a `tuple` of `Certificate`. Put another way, a single `Certificate` points only to one certificate object.